### DISCUSSION TITLE
Evaluation with mask

### DISCUSSION THREAD

**Discussion by tureckova** [obss/sahi#628] [Q&A] [answered]
Hi, if I run the evaluation script with masks, it returns the following error:
```
File "/usr/local/lib/python3.8/dist-packages/sahi/scripts/coco_evaluation.py", line 382, in evaluate
    result = evaluate_core(
  File "/usr/local/lib/python3.8/dist-packages/sahi/scripts/coco_evaluation.py", line 147, in evaluate_core
    cocoDt = cocoGt.loadRes(results)
  File "/usr/local/lib/python3.8/dist-packages/pycocotools/coco.py", line 347, in loadRes
    ann['area'] = maskUtils.area(ann['segmentation'])
  File "/usr/local/lib/python3.8/dist-packages/pycocotools/mask.py", line 95, in area
    return _mask.area(rleObjs)
  File "pycocotools/_mask.pyx", line 160, in pycocotools._mask.area
  File "pycocotools/_mask.pyx", line 128, in pycocotools._mask._frString
TypeError: list indices must be integers or slices, not str
```
Also, I noticed that in the definition of eval function is an option:
`type (bool): 'bbox' or 'mask'`
while in the definition of evaluate_core function:
`metric (str | list[str]): Metrics to be evaluated. Options are 'bbox', 'segm', 'proposal'.`

I have the last version: 0.8.22. In some previous versions, the evaluation for segmentation masks certainly worked.

Thank you for your inspection.

**Comment by fcakyon** [comment]
`type` should be `segm` instead of `mask`, i opened a pr fixing that docstring issue. Have you tried with `type=segm` ?

Can you upload a sample annotation file or a snippet of your annotation dict in json file?

**Comment by tureckova** [comment]
Yes, I gathered that `type` must be `segm`. :smile: 
Screenshot of results.json file (generated via sahi) on the left side and dataset json file on the right side:
![Screenshot from 2022-02-04 11-25-03](https://user-images.githubusercontent.com/32820891/152513310-41d88ea8-5e70-46da-b416-dc0cd5dbbf4b.png)

**Comment by tureckova** [comment]
Here are the JSON files in question
[annotation.zip](https://github.com/obss/sahi/files/8002039/annotation.zip)

**Comment by fcakyon** [comment]
@tureckova thanks a lot for providing samples, will try to look into it in 2-3 days, currently very busy with some deadlines 😅

**Comment by fcakyon** [comment]
@tureckova please uninstall `sahi` with:

```bash
pip uninstall sahi -y
```

then install `sahi` from main branch:

```bash
pip install git+github.com/obss/sahi.git@main
```

I get these results with your provided files: [eval_result.zip](https://github.com/obss/sahi/files/8053212/eval_result.zip)

```bash
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=500 ] = 0.030
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=500 ] = 0.050
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=500 ] = 0.029
 Average Precision  (AP) @[ IoU=0.50      | area= small | maxDets=500 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50      | area=medium | maxDets=500 ] = 0.053
 Average Precision  (AP) @[ IoU=0.50      | area= large | maxDets=500 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=500 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=500 ] = 0.031
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=500 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=500 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=500 ] = 0.180
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=500 ] = -1.000
```

**Comment by fcakyon** [comment]
@tureckova new version `0.9.0` is live, update your `sahi` with `pip install -U sahi` 🚀 

https://github.com/obss/sahi/releases/tag/0.9.0

**Comment by tureckova** [comment]
Hmm, I updated to version 0.9.0 and still got the same error during evaluation. Evel, if I predict the results with the new version, where might be a catch? 
Also, the results you posted seem weird to me since if I evaluate boxes only, I get mAP 0.39. Although, it might be due to changes in sahi parameters.

**Comment by fcakyon** [comment]
Sorry @tureckova, I forgot to test `segm`, my bad. I performed eval on `boxes`. Opening the issue again 👍

**Comment by claudio9russo7** [comment]
@tureckova If the images you posted refer to the same mask, I see that there is a different category_id, maybe the problem is that.

**Comment by HerongWang** [comment]
have you solve this problem? i met the same error.

**Reply by fcakyon** [reply]
This should be fixed with latest sahi release `sahi==0.11.1`

**Comment by fcakyon** [comment, marked as answer]
@tureckova I am deeply sorry for the late fix but the original issue has been fixed with the `sahi==0.11.1`

https://github.com/obss/sahi/releases/tag/0.11.1

Btw congratz on your ICIP paper! https://ieeexplore.ieee.org/abstract/document/9897826/

**Reply by tureckova** [reply]
Thank you.  My congratulations on your paper. :pray:
I was hoping to meet you there in person but didn't have such luck. Were you in Bordeaux or just online?

**Reply by fcakyon** [reply]
Yeah I was also in Bordeaux, its unfortunate we could not meet there in person 😓
